x/axelarnet/rate_limit.go

```diff
@@ -95,10 +71,42 @@ func (r RateLimiter) SendPacket(ctx sdk.Context, chanCap *capabilitytypes.Capabi
 		return nil
 	}
 
-	return r.RateLimitPacket(ctx, packet, nexus.Outgoing)
+	return r.RateLimitPacket(ctx, packet, nexus.Outgoing, types.NewIBCPath(packet.GetSourcePort(), packet.GetSourceChannel()))
 }
 
 // WriteAcknowledgement implements the ICS4 Wrapper interface
 func (r RateLimiter) WriteAcknowledgement(ctx sdk.Context, chanCap *capabilitytypes.Capability, packet ibcexported.PacketI, ack ibcexported.Acknowledgement) error {
 	return r.channel.WriteAcknowledgement(ctx, chanCap, packet, ack)
 }
+
+func parseTokenFromPacket(packet ibcexported.PacketI) (sdk.Coin, error) {
+	// Only ICS-20 packets are expected in the middleware
+	var data ibctransfertypes.FungibleTokenPacketData
+	if err := ibctransfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return sdk.Coin{}, sdkerrors.Wrapf(sdkerrors.ErrInvalidType, "cannot unmarshal ICS-20 transfer packet data: %s", err.Error())
+	}
+
+	asset := data.Denom
+	// If the asset being transferred is an IBC denom originating on the destination chain,
+	// then the full denom in the IBC transfer contains the IBC channel to the destination chain as a prefix, `transfer/channel-x/asset`.
+	// e.g. For IBC channel Axelar channel-0 <-> channel-1 Cosmoshub,
+	// For an asset `uusdc` originating on Axelar, the full denom when sending it from Cosmoshub -> Axelar will be `transfer/channel-1/uusdc`
+	// Similarly, for an asset `uatom` originating on Cosmoshub, the full denom when sending it from Axelar -> Cosmoshub will be `transfer/channel-0/uatom`
+	// So, if the source channel of the packet is a prefix of the denom being transferred, we remove it to use the remaining denom as the asset for the rate limit.
+	if ibctransfertypes.ReceiverChainIsSource(packet.GetSourcePort(), packet.GetSourceChannel(), data.Denom) {
+		ibcTransferPrefix := ibctransfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
+		asset = data.Denom[len(ibcTransferPrefix):]
```

Comment:
cool, it would solve multihop impersonate

Reply:
yep